rust/stub/src/uefi_log.rs

```diff
@@ -0,0 +1,139 @@
+//! UEFI-compatible logging backend
+
+#![deny(missing_docs)]
+#![deny(clippy::pedantic)]
+#![deny(clippy::missing_docs_in_private_items)]
+
+use alloc::{boxed::Box, string::String};
+use core::fmt::Write;
+use core::ptr::NonNull;
+use log::{Level, LevelFilter, Metadata, Record};
+use uefi::{
+    prelude::*,
+    proto::{console::text::Color, console::text::Output, loaded_image::LoadedImage},
+    Result,
+};
+
+/// Logger that logs to UEFI stdout
+struct UefiLogger {
+    /// Maximum level to log
+    max_level: LevelFilter,
+
+    /// Writer to write messages to
+    writer: Option<NonNull<Output<'static>>>,
+}
+
+impl log::Log for UefiLogger {
+    /// Returns whether or not this logger is active for a given message
+    fn enabled(&self, metadata: &Metadata<'_>) -> bool {
+        metadata.level() <= self.max_level
+    }
+
+    /// Logs a record
+    fn log(&self, record: &Record<'_>) {
+        if !self.enabled(record.metadata()) {
+            return;
+        }
+        let writer = if let Some(mut ptr) = self.writer {
+            unsafe { ptr.as_mut() }
+        } else {
+            return;
+        };
+
+        let foreground = match record.level() {
+            Level::Error => Color::Red,
+            Level::Warn => Color::Yellow,
+            Level::Info => Color::White,
+            Level::Debug => Color::Blue,
+            Level::Trace => Color::Cyan,
+        };
+        let _ = writer.set_color(foreground, Color::Black);
+        let _ = write!(writer, "{}", record.level());
+        let _ = writer.set_color(Color::White, Color::Black);
+        let _ = write!(writer, " - {}\r\n", record.args());
+    }
+
+    /// Does not do anything - needed to comply with the trait
+    fn flush(&self) {}
+}
+
+// The logger is not thread-safe, but the UEFI boot environment only uses one processor.
+unsafe impl Sync for UefiLogger {}
+unsafe impl Send for UefiLogger {}
```

Comment:
The interesting question is why it would need to be `Send + Sync` then?

Reply:
```
error[E0277]: `NonNull<Output<'static>>` cannot be sent between threads safely
    --> src/uefi_log.rs:26:6
     |
26   | impl log::Log for UefiLogger {
     |      ^^^^^^^^ `NonNull<Output<'static>>` cannot be sent between threads safely
     |
     = help: within `UefiLogger`, the trait `Send` is not implemented for `NonNull<Output<'static>>`
     = note: required because it appears within the type `Option<NonNull<Output<'static>>>`
note: required because it appears within the type `UefiLogger`
    --> src/uefi_log.rs:18:8
     |
18   | struct UefiLogger {
     |        ^^^^^^^^^^
note: required by a bound in `Log`
    --> /home/janne/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.4.17/src/lib.rs:1269:23
     |
1269 | pub trait Log: Sync + Send {
     |                       ^^^^ required by this bound in `Log`

error[E0277]: `NonNull<Output<'static>>` cannot be shared between threads safely
    --> src/uefi_log.rs:26:6
     |
26   | impl log::Log for UefiLogger {
     |      ^^^^^^^^ `NonNull<Output<'static>>` cannot be shared between threads safely
     |
     = help: within `UefiLogger`, the trait `Sync` is not implemented for `NonNull<Output<'static>>`
     = note: required because it appears within the type `Option<NonNull<Output<'static>>>`
note: required because it appears within the type `UefiLogger`
    --> src/uefi_log.rs:18:8
     |
18   | struct UefiLogger {
     |        ^^^^^^^^^^
note: required by a bound in `Log`
    --> /home/janne/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.4.17/src/lib.rs:1269:16
     |
1269 | pub trait Log: Sync + Send {
     |                ^^^^ required by this bound in `Log`

For more information about this error, try `rustc --explain E0277`.
```